--- acpid-1.0.6/Makefile.makefile	2007-06-30 21:15:45.000000000 +0200
+++ acpid-1.0.6/Makefile	2008-02-13 18:05:37.000000000 +0100
@@ -21,8 +21,9 @@ acpi_listen_OBJS = $(acpi_listen_SRCS:.c
 MAN8 = acpid.8 acpi_listen.8
 MAN8GZ = $(MAN8:.8=.8.gz)
 
-CFLAGS = -Wall -Werror -g $(DEFS)
-DEFS = -DVERSION="\"$(VERSION)\""
+CFLAGS = -Wall -Werror -g $(DEFS) $(RPM_OPT_FLAGS) -fPIE
+LDFLAGS = -pie -Wl,-z,relro
+DEFS = -DVERSION="\"$(VERSION)\"" -D_GNU_SOURCE
 
 all: $(PROGS)
 
