diff -up acpid-2.0.2/acpid.c.headers acpid-2.0.2/acpid.c
--- acpid-2.0.2/acpid.c.headers	2010-02-25 22:35:49.625058306 +0100
+++ acpid-2.0.2/acpid.c	2010-02-25 22:37:40.512063316 +0100
@@ -21,6 +21,8 @@
  */
 
 #include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
 #include <fcntl.h>
 #include <signal.h>
 #include <string.h>
diff -up acpid-2.0.2/sock.c.headers acpid-2.0.2/sock.c
--- acpid-2.0.2/sock.c.headers	2010-02-25 22:35:55.644052326 +0100
+++ acpid-2.0.2/sock.c	2010-02-25 22:37:29.592075611 +0100
@@ -21,6 +21,8 @@
  */
 
 #include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
 #include <fcntl.h>
 #include <stdio.h>
 #include <stdlib.h>
